{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "aebfacf7",
   "metadata": {},
   "source": [
    "# Splitting sequences into ORF and barcode parts\n",
    "\n",
    "After trimming the backbone sequences and reverse complementing the reverse complement sequences, all sequences have this structure:\n",
    "\n",
    "**5'-[ORF]nnn[Barcode]-3'**\n",
    "\n",
    "Where **nnn** represents the internal linker:\n",
    "\n",
    "`AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAA`\n",
    "\n",
    "Goal: split sequences into ORF and barcode parts, and add the barcode to the ORF name in the FASTQ header.\n",
    "\n",
    "Approach: use `porechop_abi` to split the sequences based on the internal linker sequence, and a custom Python script to move the barcodes to the read name in the FASTQ header of the ORF sequences."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ed42e60",
   "metadata": {},
   "source": [
    "## Testing the `porechop_abi` approach\n",
    "\n",
    "First test the `porechop_abi` approach on a small subset of the data to make sure it works as expected. Use the `cutadapt_test_04.fastq` file from the previous step, which has been optimised for trimming of backbone sequences, and contains 56.2% of reads with adapters.\n",
    "\n",
    "internal_sequence.txt:\n",
    "```\n",
    ">Internal_linker_Middle\n",
    "AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAA\n",
    "\n",
    "```\n",
    "\n",
    "Explanation of the `porechop_abi` arguments:\n",
    "- `--custom_adapters`: specifies the file containing custom adapter sequences to be used for trimming.\n",
    "- `--discard_database`: discards the internal database of adapter sequences, so only custom adapters are used.\n",
    "- `--threads 4`: uses 4 threads for parallel processing.\n",
    "- `--middle_threshold 75`: sets the minimum similarity threshold (in percentage) for detecting middle adapters.\n",
    "- `--end_threshold 95`: sets the minimum similarity threshold (in percentage) for detecting end adapters.\n",
    "- `--extra_middle_trim_good_side 0`: specifies how many base pairs to trim from the good side of middle adapters (0 means no trimming).\n",
    "- `--extra_middle_trim_bad_side 0`: specifies how many base pairs to trim from the bad side of middle adapters (0 means no trimming).\n",
    "- `--min_split_read_size 24`: sets the minimum size of a read that is considered for splitting (24 is the barcode length).\n",
    "- `--end_threshold 101`: sets the maximum allowed end threshold value (101% in this case, i.e. disable end trimming)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "adbf1955",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "DATA_DIR=\"/media/niek/4TB_SSD2/analyses/GPS_ONT\"\n",
    "porechop_abi -i $DATA_DIR/cutadapt_test_04.fastq \\\n",
    "    -o $DATA_DIR/split_trimmed_test_01.fastq \\\n",
    "    --custom_adapters internal_sequence.txt \\\n",
    "    --discard_database \\\n",
    "    --threads 4 \\\n",
    "    --middle_threshold 95 \\\n",
    "    --end_threshold 95 \\\n",
    "\t--extra_middle_trim_good_side 0 \\\n",
    "    --extra_middle_trim_bad_side 0 \\\n",
    "\t--min_split_read_size 24 \\\n",
    "\t--end_threshold 101 \\\n",
    "    > porechop_abi_test_01.log\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d25b738",
   "metadata": {},
   "source": [
    "Important info from log:\n",
    "\n",
    "```shell\n",
    "    0 / 5,621 reads had adapters trimmed from their start (0 bp removed)\n",
    "    0 / 5,621 reads had adapters trimmed from their end (0 bp removed)\n",
    "    ...\n",
    "    5,614 / 5,621 reads were split based on middle adapters\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9fdb0fe",
   "metadata": {},
   "source": [
    "5614 out of 5621 (99.89%) reads were split based on the middle adapter, which is the internal linker sequence. \n",
    "\n",
    "This is a good result, as we expect most reads to contain the internal linker sequence, and therefore be split into ORF and barcode parts.\n",
    "\n",
    "------\n",
    "\n",
    "Now run the `porechop_abi` command on the full dataset to split all sequences into ORF and barcode parts, and add the barcode to the ORF name in the FASTQ header:\n",
    "\n",
    "```shell\n",
    "#!/bin/bash\n",
    "\n",
    "#SBATCH -A JNATHAN-SL2-CPU\n",
    "#SBATCH --mail-type=BEGIN,END,FAIL\n",
    "#SBATCH -p cclake\n",
    "#SBATCH -D /home/nw416/rds/rds-jan_1-tpuFdqHBAEk/gps_ont\n",
    "#SBATCH -o ont_split.log\n",
    "#SBATCH -c 8\n",
    "#SBATCH -t 06:00:00\n",
    "#SBATCH -J ont_split\n",
    "#SBATCH --mem=60G\n",
    "\n",
    "# Initialize Conda for script usage\n",
    "source \"/home/nw416/miniforge3/etc/profile.d/conda.sh\"\n",
    "conda activate ont\n",
    "\n",
    "DATA_DIR=/home/nw416/rds/rds-jan_1-tpuFdqHBAEk/gps_ont\n",
    "TRIMMED_FASTQ=$DATA_DIR/trimmed.fastq\n",
    "SPLIT_FASTQ=$DATA_DIR/split_trimmed.fastq\n",
    "\n",
    "# Split sequences into ORF and barcode sequences \n",
    "# ---------------------------------------\n",
    "porechop_abi -i $TRIMMED_FASTQ \\\n",
    "    -o $SPLIT_FASTQ \\\n",
    "    --custom_adapters internal_sequence.txt \\\n",
    "    --discard_database \\\n",
    "    --threads 8 \\\n",
    "    --middle_threshold 95 \\\n",
    "    --end_threshold 95 \\\n",
    "\t--extra_middle_trim_good_side 0 \\\n",
    "    --extra_middle_trim_bad_side 0 \\\n",
    "\t--min_split_read_size 24 \\\n",
    "\t--end_threshold 101 \\\n",
    "    > split.log\n",
    "```\n",
    "File name: 03_split.sh\n",
    "\n",
    "------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6671f65",
   "metadata": {},
   "source": [
    "While this approach works well on a small file, it is not suitable for the full dataset, as it is extremely slow, due to the adapter trimming step that still takes place. It also takes a hugh amount of memory. Even with 60G allocated it runs out of memory and crashes after processing less than 1% of the data. This also took 30 min so we will never be able to run this on the HPC."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "156ca6f9",
   "metadata": {},
   "source": [
    "## Using `cutadapt` to split sequences into ORF and barcode parts\n",
    "\n",
    "Since the `porechop_abi` approach is not suitable for the full dataset, we will use `cutadapt` to split the sequences into ORF and barcode parts based on the internal linker sequence. \n",
    "\n",
    "This will be more efficient and faster than the `porechop_abi`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea0e3a0e",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "DATA_DIR=\"/media/niek/4TB_SSD2/analyses/GPS_ONT\"\n",
    "TRIMMED_FASTQ=\"$DATA_DIR/cutadapt_test_06.fastq\"\n",
    "\n",
    "cutadapt \\\n",
    "\t-a AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT \\\n",
    "\t-e 0.15 \\\n",
    "\t-O 40 \\\n",
    "\t--untrimmed-output $DATA_DIR/trimmed_unsplit_01.fastq \\\n",
    "\t--rest-file $DATA_DIR/barcode_01.txt \\\n",
    "\t-o $DATA_DIR/orf_01.fastq \\\n",
    "\t$TRIMMED_FASTQ > cutadapt_split_test_01.log\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "187a607c",
   "metadata": {},
   "source": [
    "Explanation of the `cutadapt` arguments and output files:\n",
    "- `-a`: specifies the 3' adapter sequence to be trimmed.\n",
    "- `-e`: sets the maximum error rate (0.15 = 15%).\n",
    "- `-O`: sets the minimum overlap length (40 nucleotides).\n",
    "- `--rest-file`: outputs sequences that came after the adapter sequence (i.e. the barcode).\n",
    "- `--untrimmed-output`: outputs sequences that do not match any adapter to FASTQ.\n",
    "- `-o`: trimmed sequences (i.e. clean ORF sequences) in FASTQ format.\n",
    "\n",
    "The format of the rest file is:\n",
    "\n",
    "```text\n",
    "TAATAGGCTGGGAGTGGTCAGTCC 00a2870a-79fb-4e1e-a961-34da5bd0a6fd runid=397513e0-9874-4c38-b5d7-80582efe53ad ch=2882 start_time=2026-02-14T16:44:08.514475-08:00 flow_cell_id=PBK43160 basecall_gpu=NVIDIA_A100_80GB_PCIe protocol_group_id=260213EUG_CUSTOM sample_id=TIMMS parent_read_id=00a2870a-79fb-4e1e-a961-34da5bd0a6fd basecall_model_version_id=dna_r10.4.1_e8.2_400bps_sup@v4.3.0 rc\n",
    "```\n",
    "\n",
    "Each line has the rest sequence (i.e. barcode) and read name separated by a space.\n",
    "\n",
    "Cut adapt log (relevant parts):\n",
    "\n",
    "```shell\n",
    "This is cutadapt 5.2 with Python 3.12.12\n",
    "Command line parameters: -a AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT -e 0.15 -O 40 --untrimmed-output /media/niek/4TB_SSD2/analyses/GPS_ONT/trimmed_unsplit_01.fastq --rest-file /media/niek/4TB_SSD2/analyses/GPS_ONT/barcode_01.fastq -o /media/niek/4TB_SSD2/analyses/GPS_ONT/orf_01.fastq /media/niek/4TB_SSD2/analyses/GPS_ONT/cutadapt_test_06.fastq\n",
    "Processing single-end reads on 1 core ...\n",
    "\n",
    "=== Summary ===\n",
    "\n",
    "Total reads processed:                   5,487\n",
    "Reads with adapters:                     5,441 (99.2%)\n",
    "\n",
    "== Read fate breakdown ==\n",
    "Reads discarded as untrimmed:               46 (0.8%)\n",
    "Reads written (passing filters):         5,441 (99.2%)\n",
    "\n",
    "Total basepairs processed:     5,621,338 bp\n",
    "Total written (filtered):      4,909,875 bp (87.3%)\n",
    "\n",
    "=== Adapter 1 ===\n",
    "\n",
    "Sequence: AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT; Type: regular 3'; Length: 96; Trimmed: 5441 times\n",
    "\n",
    "Minimum overlap: 40\n",
    "No. of allowed errors:\n",
    "1-5 bp: 0; 6-12 bp: 1; 13-19 bp: 2; 20-25 bp: 3; 26-32 bp: 4; 33-39 bp: 5; 40-45 bp: 6; 46-52 bp: 7; 53-59 bp: 8; 60-65 bp: 9; 66-72 bp: 10; 73-79 bp: 11; 80-85 bp: 12; 86-92 bp: 13; 93-96 bp: 14\n",
    "\n",
    "Bases preceding removed adapters:\n",
    "  A: 7.2%\n",
    "  C: 16.9%\n",
    "  G: 67.1%\n",
    "  T: 8.9%\n",
    "  none/other: 0.0%\n",
    "```\n",
    "\n",
    "Seems to have worked well, as nearly all reads (99.2%) had the adapter sequence and were split.\n",
    "\n",
    "One possibility is that the options were too promiscuous when matching the internal linker.\n",
    "\n",
    "Also try a more stringent approach:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "05241ee1",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "DATA_DIR=\"/media/niek/4TB_SSD2/analyses/GPS_ONT\"\n",
    "TRIMMED_FASTQ=\"$DATA_DIR/cutadapt_test_06.fastq\"\n",
    "\n",
    "cutadapt \\\n",
    "\t-a AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT \\\n",
    "\t-e 0.15 \\\n",
    "\t-O 96 \\\n",
    "\t--untrimmed-output $DATA_DIR/trimmed_unsplit_02.fastq \\\n",
    "\t--rest-file $DATA_DIR/barcode_02.txt \\\n",
    "\t-o $DATA_DIR/orf_02.fastq \\\n",
    "\t$TRIMMED_FASTQ > cutadapt_split_test_02.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "576d07d0",
   "metadata": {},
   "source": [
    "```shell\n",
    "This is cutadapt 5.2 with Python 3.12.12\n",
    "Command line parameters: -a AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT -e 0.15 -O 96 --untrimmed-output /media/niek/4TB_SSD2/analyses/GPS_ONT/trimmed_unsplit_02.fastq --rest-file /media/niek/4TB_SSD2/analyses/GPS_ONT/barcode_02.txt -o /media/niek/4TB_SSD2/analyses/GPS_ONT/orf_02.fastq /media/niek/4TB_SSD2/analyses/GPS_ONT/cutadapt_test_06.fastq\n",
    "Processing single-end reads on 1 core ...\n",
    "\n",
    "=== Summary ===\n",
    "\n",
    "Total reads processed:                   5,487\n",
    "Reads with adapters:                     5,441 (99.2%)\n",
    "\n",
    "== Read fate breakdown ==\n",
    "Reads discarded as untrimmed:               46 (0.8%)\n",
    "Reads written (passing filters):         5,441 (99.2%)\n",
    "\n",
    "Total basepairs processed:     5,621,338 bp\n",
    "Total written (filtered):      4,909,875 bp (87.3%)\n",
    "\n",
    "=== Adapter 1 ===\n",
    "\n",
    "Sequence: AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT; Type: regular 3'; Length: 96; Trimmed: 5441 times\n",
    "\n",
    "Minimum overlap: 96\n",
    "No. of allowed errors:\n",
    "1-5 bp: 0; 6-12 bp: 1; 13-19 bp: 2; 20-25 bp: 3; 26-32 bp: 4; 33-39 bp: 5; 40-45 bp: 6; 46-52 bp: 7; 53-59 bp: 8; 60-65 bp: 9; 66-72 bp: 10; 73-79 bp: 11; 80-85 bp: 12; 86-92 bp: 13; 93-96 bp: 14\n",
    "\n",
    "Bases preceding removed adapters:\n",
    "  A: 7.2%\n",
    "  C: 16.9%\n",
    "  G: 67.1%\n",
    "  T: 8.9%\n",
    "  none/other: 0.0%\n",
    "```\n",
    "\n",
    "Increasing `-O` to 96 did not change the results. This is a good sign. Try to decrease the error rate to 0.1 to see if this changes the results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fadefbc4",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "DATA_DIR=\"/media/niek/4TB_SSD2/analyses/GPS_ONT\"\n",
    "TRIMMED_FASTQ=\"$DATA_DIR/cutadapt_test_06.fastq\"\n",
    "\n",
    "cutadapt \\\n",
    "\t-a AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT \\\n",
    "\t-e 0.1 \\\n",
    "\t-O 96 \\\n",
    "\t--untrimmed-output $DATA_DIR/trimmed_unsplit_03.fastq \\\n",
    "\t--rest-file $DATA_DIR/barcode_03.txt \\\n",
    "\t-o $DATA_DIR/orf_03.fastq \\\n",
    "\t$TRIMMED_FASTQ > cutadapt_split_test_03.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc232b5e",
   "metadata": {},
   "source": [
    "```shell\n",
    "This is cutadapt 5.2 with Python 3.12.12\n",
    "Command line parameters: -a AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT -e 0.1 -O 96 --untrimmed-output /media/niek/4TB_SSD2/analyses/GPS_ONT/trimmed_unsplit_03.fastq --rest-file /media/niek/4TB_SSD2/analyses/GPS_ONT/barcode_03.txt -o /media/niek/4TB_SSD2/analyses/GPS_ONT/orf_03.fastq /media/niek/4TB_SSD2/analyses/GPS_ONT/cutadapt_test_06.fastq\n",
    "Processing single-end reads on 1 core ...\n",
    "\n",
    "=== Summary ===\n",
    "\n",
    "Total reads processed:                   5,487\n",
    "Reads with adapters:                     5,407 (98.5%)\n",
    "\n",
    "== Read fate breakdown ==\n",
    "Reads discarded as untrimmed:               80 (1.5%)\n",
    "Reads written (passing filters):         5,407 (98.5%)\n",
    "\n",
    "Total basepairs processed:     5,621,338 bp\n",
    "Total written (filtered):      4,882,040 bp (86.8%)\n",
    "\n",
    "=== Adapter 1 ===\n",
    "\n",
    "Sequence: AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT; Type: regular 3'; Length: 96; Trimmed: 5407 times\n",
    "\n",
    "Minimum overlap: 96\n",
    "No. of allowed errors:\n",
    "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50-59 bp: 5; 60-69 bp: 6; 70-79 bp: 7; 80-89 bp: 8; 90-96 bp: 9\n",
    "\n",
    "Bases preceding removed adapters:\n",
    "  A: 7.1%\n",
    "  C: 16.8%\n",
    "  G: 67.2%\n",
    "  T: 8.9%\n",
    "  none/other: 0.0%\n",
    "```\n",
    "Slight reduction in the number of reads with adapters (98.5% vs 99.2%), but still a good result. \n",
    "\n",
    "One advantage of using a more stringent error rate is that it takes less time to run, as there are fewer reads that need to be processed.\n",
    "\n",
    "Try one more time with an even more stringent error rate of 0.05:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "450d5389",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "DATA_DIR=\"/media/niek/4TB_SSD2/analyses/GPS_ONT\"\n",
    "TRIMMED_FASTQ=\"$DATA_DIR/cutadapt_test_06.fastq\"\n",
    "\n",
    "cutadapt \\\n",
    "\t-a AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT \\\n",
    "\t-e 0.05 \\\n",
    "\t-O 96 \\\n",
    "\t--untrimmed-output $DATA_DIR/trimmed_unsplit_04.fastq \\\n",
    "\t--rest-file $DATA_DIR/barcode_04.txt \\\n",
    "\t-o $DATA_DIR/orf_04.fastq \\\n",
    "\t$TRIMMED_FASTQ > cutadapt_split_test_04.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5855d02",
   "metadata": {},
   "source": [
    "```shell\n",
    "This is cutadapt 5.2 with Python 3.12.12\n",
    "Command line parameters: -a AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT -e 0.05 -O 96 --untrimmed-output /media/niek/4TB_SSD2/analyses/GPS_ONT/trimmed_unsplit_04.fastq --rest-file /media/niek/4TB_SSD2/analyses/GPS_ONT/barcode_04.txt -o /media/niek/4TB_SSD2/analyses/GPS_ONT/orf_04.fastq /media/niek/4TB_SSD2/analyses/GPS_ONT/cutadapt_test_06.fastq\n",
    "Processing single-end reads on 1 core ...\n",
    "\n",
    "=== Summary ===\n",
    "\n",
    "Total reads processed:                   5,487\n",
    "Reads with adapters:                     5,007 (91.3%)\n",
    "\n",
    "== Read fate breakdown ==\n",
    "Reads discarded as untrimmed:              480 (8.7%)\n",
    "Reads written (passing filters):         5,007 (91.3%)\n",
    "\n",
    "Total basepairs processed:     5,621,338 bp\n",
    "Total written (filtered):      4,546,224 bp (80.9%)\n",
    "\n",
    "=== Adapter 1 ===\n",
    "\n",
    "Sequence: AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT; Type: regular 3'; Length: 96; Trimmed: 5007 times\n",
    "\n",
    "Minimum overlap: 96\n",
    "No. of allowed errors:\n",
    "1-19 bp: 0; 20-39 bp: 1; 40-59 bp: 2; 60-79 bp: 3; 80-96 bp: 4\n",
    "\n",
    "Bases preceding removed adapters:\n",
    "  A: 6.6%\n",
    "  C: 16.3%\n",
    "  G: 69.1%\n",
    "  T: 8.0%\n",
    "  none/other: 0.0%\n",
    "```\n",
    "\n",
    "This is a significant reduction in the number of reads with adapters (91.3% vs 98.5%).\n",
    "\n",
    "The error rate of 0.05 translates to 4 errors allowed for a 96 bp overlap, which is still higher than the sequencing error rate (3% for Q15 and 1% for Q20). \n",
    "\n",
    "Therefore, this is a good balance between stringency and sensitivity, and we will use these parameters for the full dataset.\n",
    "\n",
    "--------\n",
    "\n",
    "SLURM script to run `cutadapt` on the full dataset:\n",
    "\n",
    "```shell\n",
    "#!/bin/bash\n",
    "\n",
    "#SBATCH -A JNATHAN-SL2-CPU\n",
    "#SBATCH --mail-type=BEGIN,END,FAIL\n",
    "#SBATCH -p cclake\n",
    "#SBATCH -D /home/nw416/rds/rds-jan_1-tpuFdqHBAEk/gps_ont\n",
    "#SBATCH -o split_slurm.log\n",
    "#SBATCH -c 8\n",
    "#SBATCH -t 01:00:00\n",
    "#SBATCH -J split\n",
    "#SBATCH --mem=16G\n",
    "\n",
    "# Initialize Conda for script usage\n",
    "source \"/home/nw416/miniforge3/etc/profile.d/conda.sh\"\n",
    "conda activate ont\n",
    "\n",
    "DATA_DIR=/home/nw416/rds/rds-jan_1-tpuFdqHBAEk/gps_ont\n",
    "TRIMMED_FASTQ=$DATA_DIR/trimmed.fastq.gz\n",
    "UNSPLIT_FASTQ=$DATA_DIR/trimmed_unsplit.fastq.gz\n",
    "BARCODES=$DATA_DIR/barcodes.txt\n",
    "ORFS=$DATA_DIR/orfs.fastq.gz\n",
    "\n",
    "cutadapt \\\n",
    "\t-a AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT \\\n",
    "\t-e 0.05 \\\n",
    "\t-O 96 \\\n",
    "\t-j 8 \\\n",
    "\t--untrimmed-output $UNSPLIT_FASTQ \\\n",
    "\t--rest-file $BARCODES \\\n",
    "\t-o $ORFS \\\n",
    "\t$TRIMMED_FASTQ > split.log\n",
    "```\n",
    "\n",
    "--------\n",
    "\n",
    "File name: 03_split.sh\n",
    "\n",
    "Relevant parts of the log:\n",
    "\n",
    "```shell\n",
    "This is cutadapt 5.2 with Python 3.12.12\n",
    "Command line parameters: -a AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT -e 0.05 -O 96 -j 8 --untrimmed-output /home/nw416/rds/rds-jan_1-tpuFdqHBAEk/gps_ont/trimmed_unsplit.fastq.gz --rest-file /home/nw416/rds/rds-jan_1-tpuFdqHBAEk/gps_ont/barcodes.txt -o /home/nw416/rds/rds-jan_1-tpuFdqHBAEk/gps_ont/orfs.fastq.gz /home/nw416/rds/rds-jan_1-tpuFdqHBAEk/gps_ont/trimmed.fastq.gz\n",
    "Processing single-end reads on 8 cores ...\n",
    "\n",
    "=== Summary ===\n",
    "\n",
    "Total reads processed:              13,329,423\n",
    "Reads with adapters:                12,262,199 (92.0%)\n",
    "\n",
    "== Read fate breakdown ==\n",
    "Reads discarded as untrimmed:        1,067,224 (8.0%)\n",
    "Reads written (passing filters):    12,262,199 (92.0%)\n",
    "\n",
    "Total basepairs processed: 13,622,666,905 bp\n",
    "Total written (filtered):  11,026,528,420 bp (80.9%)\n",
    "\n",
    "=== Adapter 1 ===\n",
    "\n",
    "Sequence: AACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGT; Type: regular 3'; Length: 96; Trimmed: 12262199 times\n",
    "\n",
    "Minimum overlap: 96\n",
    "No. of allowed errors:\n",
    "1-19 bp: 0; 20-39 bp: 1; 40-59 bp: 2; 60-79 bp: 3; 80-96 bp: 4\n",
    "\n",
    "Bases preceding removed adapters:\n",
    "  A: 7.0%\n",
    "  C: 15.2%\n",
    "  G: 70.7%\n",
    "  T: 7.1%\n",
    "  none/other: 0.0%\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc660704",
   "metadata": {},
   "source": [
    "The output of the rest file only contains the barcode sequence and read name. It will be moved to the FASTQ header of the ORF sequences in the next step. \n",
    "\n",
    "The drawback of the rest file is that it does not contain the quality scores of the barcode sequences, which could be useful later for potential troubleshooting.\n",
    "\n",
    "Use the following Python script to convert the rest file into FASTQ format:\n",
    "\n",
    "--------\n",
    "\n",
    "```python\n",
    "import sys\n",
    "import gzip\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Usage: python extract_barcodes.py barcodes.txt trimmed.fastq.gz output_barcodes.fastq.gz\n",
    "\n",
    "rest_file = sys.argv[1]\n",
    "original_fastq = sys.argv[2]\n",
    "output_file = sys.argv[3]\n",
    "\n",
    "# Map Read ID -> Barcode Sequence\n",
    "barcode_map = {}\n",
    "with open(rest_file, \"r\") as f:\n",
    "    for line in f:\n",
    "        parts = line.strip().split(maxsplit=1)\n",
    "        if len(parts) == 2:\n",
    "            seq, header = parts\n",
    "            read_id = header.split()[0]  # Get the UUID\n",
    "            barcode_map[read_id] = seq\n",
    "\n",
    "# Extract FASTQ records\n",
    "open_func = gzip.open if original_fastq.endswith(\".gz\") else open\n",
    "out_func = gzip.open if output_file.endswith(\".gz\") else open\n",
    "\n",
    "with open_func(original_fastq, \"rt\") as in_handle, out_func(\n",
    "    output_file, \"wt\"\n",
    ") as out_handle:\n",
    "    for record in SeqIO.parse(in_handle, \"fastq\"):\n",
    "        if record.id in barcode_map:\n",
    "            target_seq = barcode_map[record.id]\n",
    "            # Find where the barcode starts in the original read\n",
    "            # We search from the end to be safe with ONT noise\n",
    "            start_pos = record.seq.rfind(target_seq)\n",
    "\n",
    "            if start_pos != -1:\n",
    "                # Slice the record (this automatically slices quality scores too!)\n",
    "                barcode_record = record[start_pos:]\n",
    "                SeqIO.write(barcode_record, out_handle, \"fastq\")\n",
    "\n",
    "print(f\"Done! Processed {len(barcode_map)} barcodes.\")\n",
    "```\n",
    "\n",
    "--------\n",
    "\n",
    "File name: extract_barcodes.py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e165f307",
   "metadata": {},
   "source": [
    "First test with the small subset of the data generated earlier to make sure it works as expected:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "92bfd9a6",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done! Processed 5006 barcodes.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "DATA_DIR=\"/media/niek/4TB_SSD2/analyses/GPS_ONT\"\n",
    "REST_FILE=$DATA_DIR/barcode_04.txt\n",
    "TRIMMED_FASTQ=\"$DATA_DIR/cutadapt_test_06.fastq\"\n",
    "OUTPUT_FILE=\"$DATA_DIR/barcode_04_extracted.fastq\"\n",
    "\n",
    "python extract_barcodes.py $REST_FILE $TRIMMED_FASTQ $OUTPUT_FILE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "041f3678",
   "metadata": {},
   "source": [
    "Check first line of the rest file:\n",
    "```\n",
    "TAATAGGCTGGGAGTGGTCAGTCC 00a2870a-79fb-4e1e-a961-34da5bd0a6fd runid=397513e0-9874-4c38-b5d7-80582efe53ad ch=2882 start_time=2026-02-14T16:44:08.514475-08:00 flow_cell_id=PBK43160 basecall_gpu=NVIDIA_A100_80GB_PCIe protocol_group_id=260213EUG_CUSTOM sample_id=TIMMS parent_read_id=00a2870a-79fb-4e1e-a961-34da5bd0a6fd basecall_model_version_id=dna_r10.4.1_e8.2_400bps_sup@v4.3.0 rc\n",
    "```\n",
    "\n",
    "Check first record in barcode FASTQ file:\n",
    "```\n",
    "@00a2870a-79fb-4e1e-a961-34da5bd0a6fd runid=397513e0-9874-4c38-b5d7-80582efe53ad ch=2882 start_time=2026-02-14T16:44:08.514475-08:00 flow_cell_id=PBK43160 basecall_gpu=NVIDIA_A100_80GB_PCIe protocol_group_id=260213EUG_CUSTOM sample_id=TIMMS parent_read_id=00a2870a-79fb-4e1e-a961-34da5bd0a6fd basecall_model_version_id=dna_r10.4.1_e8.2_400bps_sup@v4.3.0 rc\n",
    "TAATAGGCTGGGAGTGGTCAGTCC\n",
    "+\n",
    "JGNILKSJ0DCEJGFKGJNJGHHI\n",
    "```\n",
    "\n",
    "Check first record in trimmed FASTQ file:\n",
    "```\n",
    "@00a2870a-79fb-4e1e-a961-34da5bd0a6fd runid=397513e0-9874-4c38-b5d7-80582efe53ad ch=2882 start_time=2026-02-14T16:44:08.514475-08:00 flow_cell_id=PBK43160 basecall_gpu=NVIDIA_A100_80GB_PCIe protocol_group_id=260213EUG_CUSTOM sample_id=TIMMS parent_read_id=00a2870a-79fb-4e1e-a961-34da5bd0a6fd basecall_model_version_id=dna_r10.4.1_e8.2_400bps_sup@v4.3.0 rc\n",
    "GAATATCATCCTGATTTAGAAAATTTGGATGAAGATGGATATACTCAATTACACTTCGACTCTCAAAGCAATACCAGGATAGCTGTTGTTTCAGAGAAAGGATTGTGTGCTGCATCTCCTCCTTGGCGCCTCATTGCTGTAATTTTGGGAATCCTATGCTTGGTAATACTGGTGATAGCTGTGGTCCTGGGTACCATGGGGGTTCTTTCCAGCCCTTGTCCTCCTAATTGGATTATATATGAGAAGAGCTGTTATCTATTCAGCATGTCACTAAATTCCTGGGATGGAAGTAAAAGACAATGCTGGCAACTGGGCTCTAATCTCCTAAAGATAGACAGCTCAAATGAATTGGTAAGTGTAGACTTCTGTTATGATTATCTGTGGTGTGTATCCTAGAACCCAGCTTTCTTGTACAAAGTGGTTTAATGAGTTTAAACCTCGAGCGTAACTATAACGGTCCTAAGGTAGCGAACCAGTAGGTCCACTATGAGTTAATAGGCTGGGAGTGGTCAGTCC\n",
    "+\n",
    "KGJHHGFCDDDESHKIOSRPFFEEGFHGSISIKJPKNKFFIDEFFIJMHGJGHIJGHJIKJIMHKPGGGGGEEFFGFFEDFEGGGHLPRONIKSMGNHH(((((BFGCCEDSLHJHKILHHHGHGIIPSSPQNJJOISRKROLIFBJHKNMMMLLSNINKIJNMKQLHSJIKKNKJNJJNLMMKRPGFHICGNSLSMSNHJJSPQNOQIMMKOSMSPLJGKKKSGGFHHJGJGLHPILKQQNNSSJJQMKSLSOKJKIJGHMJKSLLNOLPPNLNKJEHDHIJJNNPSJSOSSNJNNKMFGF88876@DDCGGIHGGJGMSSJLLHSMSPOLOLPIHNOKKJJQNSMIKKJIQIOIILHIKOLJLISKKQHIIPNMOISOSQNLPSLSLIHIMOSNLSIHHMEACBAFFEKJLISNMKLLNRSKSSKSNOSSSROJIHH@@@@AMJKLPQSHHHFGG@@=>FOHHHGEFIKGHIKIJJJMPIGJJHSLIHJGJGNILKSJ0DCEJGFKGJNJGHHI\n",
    "```\n",
    "\n",
    "Sequence and scores match between barcode FASTQ and trimmed FASTQ files, which means the script works as expected."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cfab787",
   "metadata": {},
   "source": [
    "Now run on HPC with full dataset:\n",
    "\n",
    "--------\n",
    "\n",
    "```shell\n",
    "#!/bin/bash\n",
    "\n",
    "#SBATCH -A JNATHAN-SL2-CPU\n",
    "#SBATCH --mail-type=BEGIN,END,FAIL\n",
    "#SBATCH -p cclake\n",
    "#SBATCH -D /home/nw416/rds/rds-jan_1-tpuFdqHBAEk/gps_ont\n",
    "#SBATCH -o barcode_fq.log\n",
    "#SBATCH -c 4\n",
    "#SBATCH -t 02:00:00\n",
    "#SBATCH -J barcode_fq\n",
    "#SBATCH --mem=8G\n",
    "\n",
    "# Initialize Conda for script usage\n",
    "source \"/home/nw416/miniforge3/etc/profile.d/conda.sh\"\n",
    "conda activate ont\n",
    "\n",
    "DATA_DIR=/home/nw416/rds/rds-jan_1-tpuFdqHBAEk/gps_ont\n",
    "TRIMMED_FASTQ=$DATA_DIR/trimmed.fastq.gz\n",
    "BARCODES=$DATA_DIR/barcodes.txt\n",
    "\n",
    "python extract_barcodes.py $BARCODES $TRIMMED_FASTQ $DATA_DIR/barcodes.fastq.gz\n",
    "```\n",
    "\n",
    "--------\n",
    "\n",
    "File name: 03_barcode_fastq.sh"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ont",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
